lib/guest_installation_and_configuration_base.pm

```diff
@@ -0,0 +1,1581 @@
+        }
+        elsif ((get_var('REPO_0_TO_INSTALL', '') ne '') and ($self->{guest_version} eq lc get_required_var('TARGET_DEVELOPING_VERSION'))) {
+            $self->{guest_build} = lc get_required_var('BUILD');
+        }
+        else {
+            $self->{guest_build} = 'gm';
+        }
+        record_info("Guest $self->{guest_name} does not have guest_build set.Set it to test suite setting BUILD or GM according to guest_version", "Please pay attention ! It is now $self->{guest_build}");
+    }
+    return $self;
+}
+
+#Print out guest instance parameters
+sub print_guest_params {
+    my $self = shift;
+
+    $self->reveal_myself;
+    diag("The guest instance $self->{guest_name} created is as below:");
+    foreach (keys %{$self}) {
+        if (defined $self->{$_}) {
+            if (ref($self->{$_}) eq 'HASH') {
+                diag $_ . ' => ' . Dumper(\%{$self->{$_}});
+            }
+            elsif (ref($self->{$_}) eq 'ARRAY') {
+                diag $_ . ' => ' . @{$self->{$_}};
+            }
+            else {
+                diag $_ . ' => ' . $self->{$_};
+            }
+        }
+        else {
+            diag $_ . ' => ' . 'undef';
+        }
+    }
+    return $self;
+}
+
+#Install necessary packages, patterns, setup ssh config, create [common_log_folder].These are common environment affairs which will be used for all guest instances.
+sub prepare_common_environment {
+    my $self = shift;
+
+    $self->reveal_myself;
+    if ($common_environment_prepared eq 'false') {
+        $self->clean_up_all_guests;
+        script_run("rm -f -r $common_log_folder");
+        assert_script_run("mkdir -p $common_log_folder");
+        my @stuff_to_backup = ('/root/.ssh/config');
+        virt_autotest::utils::backup_file(\@stuff_to_backup);
+        script_run("rm -f -r /root/.ssh/config");
+        virt_autotest::utils::setup_common_ssh_config('/root/.ssh/config');
+        enable_debug_logging;
+        my $_packages_to_check = 'wget curl screen dnsmasq xmlstarlet yast2-schema python3 nmap';
+        zypper_call("install -y $_packages_to_check");
+        my $_patterns_to_check = 'kvm_server kvm_tools';
+        $_patterns_to_check = 'xen_server xen_tools' if ($self->{host_virt_type} eq 'xen');
+        zypper_call("install -y -t pattern $_patterns_to_check");
+        $common_environment_prepared = 'true';
+        diag("Common environment preparation is done now.");
+    }
+    else {
+        diag("Common environment preparation had already been done.");
+    }
+    return $self;
+}
+
+#Remove all existing guests and storage files in /var/lib/libvirt/images/
+sub clean_up_all_guests {
+    my $self = shift;
+
+    $self->reveal_myself;
+    my @_guests_to_clean_up = split(/\n/, script_output("virsh list --all --name", proceed_on_failure => 1));
+    if (scalar(@_guests_to_clean_up) gt 0) {
+        diag("Going to clean up all guests on $self->{host_name}");
+        foreach (@_guests_to_clean_up) {
```

Comment:
would be better to use assert_script_run() to clean up all guests on vm host,  instead of using script_run()  under `clean_up_all_guests()`.

Reply:
Better now ?

Reply:
Would be better:
```
- script_run("virsh destroy $_");
+ assert_script_run("virsh destroy $_") 
```
imho, there was the product issue to destroy guest.

Reply:
What do you mean ?

Reply:
yes,  I totally agree on what you said here. but sometime `assert_script_run()` will help us figure out new product failure from some important test step - such as virsh start vm or virsh destroy vm and so on. so, I just recommend to use with assert_script_run() without other ideas here.

Reply:
clean_up_all_guests and power_cycle_guest are generic subroutines. They can be used for the purpose indicated by their names, but also more importantly here to facilitate guest installation. Doing too much assertion in them can lead to unncessary false failures because of unknown guest state. assert_script_run can be used here extensively.

Reply:
yes, totally make sense now. keep using script_run("virsh destroy $_"); here.